src/octoprint/util/comprehensive_json.py

```diff
@@ -0,0 +1,141 @@
+__license__ = "GNU Affero General Public License http://www.gnu.org/licenses/agpl.html"
+__copyright__ = "Copyright (C) 2018 The OctoPrint Project - Released under terms of the AGPLv3 License"
+
+import json
+import time
+import typing
+from collections import OrderedDict
+from typing import Any, Callable, TypeVar
+
+from frozendict import frozendict
+
+from octoprint.util import to_unicode
+
+
+def dumps(obj: Any) -> str:
+    """
+    Dump an object to JSON, handles additional types that the JSON encoder can't, like
+    datetime.struct_time, bytes, and frozendicts.
+
+    Cannot handle circular references.
+
+    Note: at the moment, datetime.struct_time are only supported if they are in dicts
+    """
+    return json.dumps(
+        JsonEncoding.encode(obj),
+        separators=(",", ":"),
+        indent=None,
+        allow_nan=False,
+    )
+
+
+def loads(s: str) -> Any:
+    return json.loads(s, object_hook=JsonEncoding.decode)
+
+
+T = TypeVar("T")
+
+
+class JsonEncoding:
+    decoders: OrderedDict[str, Callable[[dict], object]] = OrderedDict()
+    encoders: OrderedDict[type, Callable[[Any], Any]] = OrderedDict()
+
+    @classmethod
+    def add_encoder(cls, typ: typing.Type[T], encoder: Callable[[T], Any]) -> None:
+        """
+        Add an encoder for a type.
+
+        :param typ: the type to add an encoder for
+        :param encoder: the encoder. Must take a single argument and return a
+            tuple (name, parameters...)
+        """
+        cls.encoders[typ] = encoder
+
+    @classmethod
+    def remove_encoder(cls, typ):
+        try:
+            del cls.encoders[typ]
+        except KeyError:
+            pass
+
+    @classmethod
+    def add_decoder(cls, classname, decoder):
+        cls.decoders[classname] = decoder
+
+    @classmethod
+    def remove_decoder(cls, classname):
+        try:
+            del cls.decoders[classname]
+        except KeyError:
+            pass
+
+    @classmethod
+    def encode(cls, val):
```

Comment:
this may be a bit overkill, but I've basically replicated the way that the json encoder walks over the value tree here, with the exception of the cycle checking that it does.

Reply:
Felt like overkill, but I couldn't find a better approach either.